### DISCUSSION TITLE
How to use ts-node esm loader with npx?

### DISCUSSION THREAD

**Discussion by Maxim-Mazurok** [TypeStrong/ts-node#1463] [Q&A] [answered]
A follow-up to https://github.com/TypeStrong/ts-node/issues/1007#issuecomment-921541774 question:

Is there any way to use ts-node esm loader with `npx`?
I was using this: `npx -y ts-node --project ./tools/tsconfig.json ./tools/not-ignored-files.ts`
Now I'm stuck with `npx -y cross-env TS_NODE_PROJECT=\"tools/tsconfig.json\" node --loader ts-node/esm --experimental-specifier-resolution=node` which requires devs to install project deps before running the script.

Perhaps, you could make `npx -y ts-node/esm ...` available?

Or is there any way I can specify something like `--loader npx:ts-node/esm`?...

`npx --package=ts-node --call "node --loader ts-node/esm --experimental-specifier-resolution=node ./tools/not-ignored-files.ts"` also didn't work for me.

**Comment by cspotcode** [comment, marked as answer]
I think the problem is that npx will install and run an executable into a non-local location controlled by npx, but the working directory is local.  Node's `--loader` flag attempts to `import` the loader, which will be done locally unless it is an absolute path.

This is why the following:

```
npx --package ts-node node --loader ts-node/esm ./example.ts
```

Will raise an error `Cannot find package 'ts-node'`.  `ts-node` cannot be found in the working directory.

We have a ticket for this: #1258

That ticket proposes the addition of a new executable: `ts-node-esm` which can be used as follows:

```
npx --package ts-node ts-node-esm ./example.ts
```

**Reply by Maxim-Mazurok** [reply]
Good point, looks like I can do something like this if I really want to:
```
npx --package ts-node@10.2.1 --call "node --loader ~/.npm/_npx/1bf7c3c15bf47d04/node_modules/ts-node/esm.mjs ...
```

But I probably will wait for that ticket to be closed.

I was hoping that there's some smart way to run node in the "npx context" or something...

**Reply by cspotcode** [reply]
Yeah, I guess the problem is you either have the wrong working directory for `--loader ts-node/esm` or the wrong working directory for your script, `./example.ts`

For what it's worth, that ticket is one of our simpler ones, which is why I have been focused on other tickets instead, hoping someone will send a pull request.  It seems like many people want it but not enough to implement it.

**Comment by sastraxi** [comment]
For what it's worth, I experienced the same error trying to use `node --loader ts-node/esm` with `"type": "module"` in a yarn 3 workspace. I also had `"ts-node": "^10.4.0"` in `devDependencies`.

> `Cannot find package 'ts-node'`

For my particular solution, `tsc-watch --onSuccess="node ..."` was a viable strategy, so I never figured out how to solve it, sadly. Just putting my results here hopefully to reduce some frustration for any other dev coming up against the same.

**Reply by Maxim-Mazurok** [reply]
Was `npx` involved? This doesn't seem to be related otherwise...

**Reply by sastraxi** [reply]
It's not directly related, but yarn versions 2 and above do very different things with package resolution, and my hypothesis is that the underlying mechanism is similar. This is the only relevant result that comes up from a google search, so I've put it here to help any future spelunkers.
